File: Data Structures & Algorithms/subsets/submission-5.java
class Solution {
    public List<List<Integer>> result = new ArrayList<>();

    public List<List<Integer>> subsets(int[] nums) {
        backtrack (new ArrayList<>(), nums, 0);

        return result;
    }

    private void backtrack (List<Integer> list, int[] nums, int start) {
        result.add (new ArrayList<>(list));

        for (int i = start; i < nums.length; i++) {
            list.add (nums[i]);
            backtrack (list, nums, i + 1);
            list.remove (list.size() - 1);
        }
    }
}

File: Data Structures & Algorithms/subsets-ii/submission-5.cpp
class Solution {
public:
    vector<vector<int>> res;

    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        backtrack(nums, {}, 0);
        return res;
    }

    void backtrack(vector<int>& nums, vector<int> subset, int start) {
        res.push_back(subset);
        for(int i = start; i < nums.size(); i++) {
            if(i > start && nums[i] == nums[i - 1]) continue;
            subset.push_back(nums[i]);
            backtrack(nums, subset, i + 1);
            subset.pop_back();
        }
    }
};
